## KnowEnGSpreadsheetOrientedObject:   KSOO v00
    * Introspective Spreadsheet in the context of KnowEnG Pipeline Anaylsis
    * HBox Directory - File Selector

In [3]:
import os
from os.path import expanduser

from IPython.display import display
from ipywidgets import widgets

file_select_box_layout = widgets.Layout(display='inline-flex', 
                                        flex_flow='row',
                                        justify_content='space-between', 
                                        align_items='stretch',
                                        border='none', width='100%')

PARAMETER_FILE_TYPES = ['.yml']
USER_DATAFILE_EXTENSIONS_LIST = ['.tsv', '.txt', '.df', '.gz', '.csv']

def user_data_list(target_dir, FEXT):
    """ user_file_list = update_user_data_list(user_data_dir, FEXT)
    Args:
        target_dir:     directory to list
        FEXT:           File extension list e.g. ['.tsv', '.txt']
    """
    my_file_list = []
    for f in os.listdir(target_dir):
        if os.path.isfile(os.path.join(target_dir, f)):
            noNeed, f_ext = os.path.splitext(f)
            if f_ext in FEXT:
                my_file_list.append(f)
            else:
                print(f)

    if len(my_file_list) <= 0:
        my_file_list.append('No Data')

    return my_file_list

def get_dirs_below(dir_in):
    """ Danger Recursion: get the dictionary of directories below this one """
    if os.path.isdir(dir_in) == False:
        return {}
    
    ignores_list = ['.', '_']
    dir_dict = {}
    for mbd in os.listdir(dir_in):
        if mbd[0] in ignores_list: continue
        dir_in_q = os.path.join(dir_in, mbd)
        if os.path.isdir(dir_in_q):
            dir_dict[mbd] = dir_in_q
            dir_below_dict = get_dirs_below(dir_in_q)
            if len(dir_below_dict) > 0:
                for d, fpd in dir_below_dict.items():
                    dir_dict[d] = fpd    
    return dir_dict

class HorizontalFileSelector():
    def __init__(self, user_home_dir=None):
        if user_home_dir is None or os.path.isdir(os.path.abspath(user_home_dir)) == False:
            self._user_data_directory = expanduser('~')
        else:
            self._user_data_directory = os.path.abspath(user_home_dir)
        
        #                             Create the set of controls in a horizontal Box
        self.directory_valid = widgets.Valid(value=False, description='directory')
        self.directory_dropd = widgets.Dropdown(options=['none'], value='none', description='', disabled=True)

        self.filename_valid  = widgets.Valid(value=False, description='file')
        self.filename_dropd  = widgets.Dropdown(options=[' ','none'], value=' ', description='', disabled=True)
        self.select_button = widgets.Button(description='Select', disabled=True)

        self.file_select_set = widgets.HBox([self.directory_valid, 
                                             self.directory_dropd,
                                             self.filename_valid, 
                                             self.filename_dropd, 
                                             self.select_button], 
                                            layout=file_select_box_layout)
        self.set_directory_dropd()
        self.set_filename_dropd('null')
        
        self.directory_dropd.observe(self.set_filename_dropd, 'value')
        
    def show(self):
        display(self.file_select_set)
        
    def get_directory_dropd(self):
        return os.path.join(self._user_data_directory, self.directory_dropd.value)
        
    def set_directory_dropd(self):
        dropd_dir_dict = get_dirs_below( self._user_data_directory )
        if len(dropd_dir_dict) > 0:
            dropd_dir_list = sorted(list(dropd_dir_dict.keys()))
            self.directory_dropd.options = dropd_dir_list
            self.directory_dropd.value   = dropd_dir_list[0]
            selected_directory = self.get_directory_dropd()
            
            if os.path.isdir(selected_directory):
                self.set_filename_dropd('null')
            self.directory_dropd.disabled = False
            self.directory_valid.value = True
        else:
            self.directory_dropd.options = ['none']
            self.directory_dropd.value   = 'none'
            self.directory_dropd.disabled = True
            self.directory_valid.value = False

    def set_filename_dropd(self, _):
        selected_dir = os.path.join(self._user_data_directory, self.directory_dropd.value)
        self.filename_dropd.options=user_data_list(selected_dir, USER_DATAFILE_EXTENSIONS_LIST)
        self.filename_dropd.value = self.filename_dropd.options[0]
        self.filename_dropd.disabled = False
        self.filename_valid.value = True
        self.select_button.disabled = False
        

In [5]:
home = '/Users/mojo/git_clone/dlanier/Spreadsheets_Transformation/'

# os.path.isdir(home)
hfs = HorizontalFileSelector('home)
hfs.show()

user_home_dir home False


FileNotFoundError: [Errno 2] No such file or directory: '/Users/mojo/       536-dynamic.lm'

```
    user_home_dir home False
    ---------------------------------------------------------------------------
    FileNotFoundError                         Traceback (most recent call last)
    <ipython-input-5-73cfcc4df7de> in <module>()
          2 
          3 # os.path.isdir(home)
    ----> 4 hfs = HorizontalFileSelector('home')
          5 hfs.show()

    <ipython-input-3-0666eda59904> in __init__(self, user_home_dir)
         77                                             layout=file_select_box_layout)
         78         self.set_directory_dropd()
    ---> 79         self.set_filename_dropd('null')
         80 
         81         self.directory_dropd.observe(self.set_filename_dropd, 'value')

    <ipython-input-3-0666eda59904> in set_filename_dropd(self, _)
        108     def set_filename_dropd(self, _):
        109         selected_dir = os.path.join(self._user_data_directory, self.directory_dropd.value)
    --> 110         self.filename_dropd.options=user_data_list(selected_dir, USER_DATAFILE_EXTENSIONS_LIST)
        111         self.filename_dropd.value = self.filename_dropd.options[0]
        112         self.filename_dropd.disabled = False

    <ipython-input-3-0666eda59904> in user_data_list(target_dir, FEXT)
         22     """
         23     my_file_list = []
    ---> 24     for f in os.listdir(target_dir):
         25         if os.path.isfile(os.path.join(target_dir, f)):
         26             noNeed, f_ext = os.path.splitext(f)

    FileNotFoundError: [Errno 2] No such file or directory: '/Users/mojo/       536-dynamic.lm'
'''